src/java/org/apache/cassandra/index/sai/utils/OrderingFilterRangeIterator.java

```diff
@@ -19,76 +19,57 @@
 package org.apache.cassandra.index.sai.utils;
 
 import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Set;
 import java.util.function.Function;
 import javax.annotation.concurrent.NotThreadSafe;
 
+import org.apache.cassandra.index.sai.SSTableIndex;
 import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.CloseableIterator;
+import org.apache.cassandra.utils.Pair;
 
 /**
  * An iterator that consumes a chunk of {@link PrimaryKey}s from the {@link RangeIterator}, passes them to the
  * {@link Function} to filter the chunk of {@link PrimaryKey}s and then pass the results to next consumer.
  * The PKs are currently returned in {@link PrimaryKey} order, but that contract may change.
  */
 @NotThreadSafe
-public class OrderingFilterRangeIterator extends RangeIterator
+public class OrderingFilterRangeIterator implements Iterator<Pair<List<CloseableIterator<ScoredPrimaryKey>>, Set<SSTableIndex>>>, AutoCloseable
```

Comment:
My rule of thumb is, it's fine to use Pair in private methods but public ones should declare a class with named fields and I think this is definitely complex enough that it shouldn't be an exception :)

Reply:
Yeah, this is pretty egregious and definitely doesn't warrant an exception :)

Reply:
I added a class. I still think it might be worth asking whether we need `OrderingFilterRangeIterator`. Is there a point where we materialize enough PrimaryKeys for a where clause that we decide to sort then filter?

Reply:
Isn't this what Piotr's code around `QueryController.decideFilterSortOrder` does?

Reply:
We'll leave it as is. I added a metric that we can see on query traces.